BUG: Preserve Series/DataFrame subclasses through groupby operations

JBGreisman opened the pull request:
This pull request fixes a bug that caused subclassed Series/DataFrames to revert to `pd.Series` and `pd.DataFrame` after `groupby()` operations. This is a follow-up on a previously abandoned pull request (#28573)

- [x] closes #28330
- [x] tests added / passed
- [x] passes black pandas
- [x] passes git diff upstream/master -u -- "*.py" | flake8 --diff
- [x] whatsnew entry

JBGreisman committed:
Add tests to confirm groupby ops preserve subclasses

JBGreisman committed:
Update SeriesGroupBy constructor calls to preserve subclassed Series

JBGreisman committed:
Fix concat() method to preserve subclassed DataFrames

JBGreisman committed:
Add GH28330 comment to concat.py

JBGreisman committed:
Preserve subclassing in DataFrame.idxmin() and DataFrame.idxmax() calls

JBGreisman committed:
GH28330 Fix GroupBy.size() to preserve subclassed types

JBGreisman committed:
GH28330 Fix GroupBy.ngroup() to preserve subclassed types

JBGreisman committed:
GH28330 Fix GroupBy.cumcount() to preserve subclassed types

JBGreisman committed:
GH28330 Fix constructor calls to preserve subclasses through groupby()

JBGreisman committed:
Fix typo -- Series.constructor() to Series._constructor()

JBGreisman committed:
Remove DeprecationWarning due to empty Series construction

JBGreisman committed:
BUG: GH28330 Preserve subclassing with groupby operations

JBGreisman committed:
BUG: GH28330 Preserve subclassing with groupby operations

JBGreisman committed:
Merge remote-tracking branch 'upstream/master' into groupby-preserve-subclass

JBGreisman committed:
Fix formatting of .py files with black

JBGreisman committed:
Removed trailing whitespace in doc/source/whatsnew/v1.1.0.rst

JBGreisman committed:
Update DataFrameGroupBy._cython_agg_blocks() to pass mypy

JBGreisman committed:
Merge remote-tracking branch 'upstream/master' into groupby-preserve-subclass

JBGreisman committed:
Remove unused import of typing.cast from pandas/core/groupby/generic.py

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -701,6 +701,7 @@ Groupby/resample/rolling
 - Bug in :meth:`DataFrame.groupby` where a ``ValueError`` would be raised when grouping by a categorical column with read-only categories and ``sort=False`` (:issue:`33410`)
 - Bug in :meth:`GroupBy.first` and :meth:`GroupBy.last` where None is not preserved in object dtype (:issue:`32800`)
 - Bug in :meth:`Rolling.min` and :meth:`Rolling.max`: Growing memory usage after multiple calls when using a fixed window (:issue:`30726`)
+- Bug where subclasses were not preserved through groupby operations (:issue:`28330`)
```
can you enumerate which operations here

jreback replied:
you can also do general classes of things, e.g. fitler, transform, agg if it affects lots of functions.

jreback commented on pandas/core/groupby/groupby.py:
```diff
@@ -2047,7 +2051,13 @@ def ngroup(self, ascending: bool = True):
             index = self._selected_obj.index
-            result = Series(self.grouper.group_info[0], index)
+
+            # GH28330 preserve subclassed Series/DataFrames through calls
```
why is this needed?

JBGreisman replied:
Returning the result of `GroupBy.ngroup()` as a `Series` ends up erasing the subclass information from `result._constructor_expanddim()`, which led to the default return of a `DataFrame`/`Series` instead of the subclass. 

It seemed to me that `self.obj` was either a `Series` or `DataFrame` instance (or subclass of those), depending on if was called by a `SeriesGroupBy` or `DataFrameGroupBy` so the if statement ensures that the proper `Series` (or subclass) is returned.

jreback commented on pandas/core/groupby/groupby.py:
```diff
@@ -2109,7 +2119,12 @@ def cumcount(self, ascending: bool = True):
             cumcounts = self._cumcount_array(ascending=ascending)
-            return Series(cumcounts, index)
+
+            # GH28330 preserve subclassed Series/DataFrames through calls
```
why is this needed?

JBGreisman replied:
This was the same logic as for `GroupBy.ngroup()`

jreback replied:
I would actually leave these. I don't think returning a subclass makes any sense here.

JBGreisman replied:
I reverted these changes -- `ngroup()` and `cumcount()` -- in the most recent set of commits.

JBGreisman replied:
Thinking about this a bit more -- I'm not sure if it makes sense to have certain groupby-agg operations return the parent `Series` rather than a subclass. It seems more consistent to me to have all groupby.agg and groupby.transform preserve subclasses.

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -195,6 +195,42 @@ def f(grp):
     tm.assert_series_equal(result, e)
 
 
+@pytest.mark.parametrize(
```
make a new file test_groupby_subclass.

JBGreisman replied:
will do

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -195,6 +195,42 @@ def f(grp):
+        tm.SubclassedSeries(np.arange(0, 10), name="A"),
+    ],
+)
+def test_groupby_preserves_subclass(obj, groupby_func):
```
does this cover *all* of the changes? e.g. I see some that are in non-groupby files, please add tests for those as well.

JBGreisman replied:
To my knowledge, it does hit almost all of the changes in other files (I'll add any needed tests after combing through the codecov report).  For example:
- the changes in `frame.py` end up being hit by the `groupby.idxmin()` and `groupby.idxmax()` tests
- the changes to `concat.py` end up being hit by the `groupby.transform()` operations via `DataFrameGroupBy.__transform_general()`

I will add more targeted tests though to make sure other ways of calling those functions still preserve subclassing.

JBGreisman committed:
Move tests to test_groupby_subclass.py

JBGreisman committed:
Add tests for DataFrame.idxmin() and DataFrame.idxmax() with subclasses

JBGreisman committed:
Add test to confirm concat() preserves subclassed types

JBGreisman committed:
Update whatsnew entry bugfix

JBGreisman committed:
Merge remote-tracking branch 'upstream/master' into groupby-preserve-subclass

JBGreisman commented:
I updated my branch with additional tests, and a revised entry in the whatsnew rst file. 

Please let me know if there are any additional recommendations for how to proceed with this PR.

jreback commented on pandas/core/groupby/generic.py:
```diff
@@ -1101,7 +1108,8 @@ def _cython_agg_blocks(
                 else:
-                    result = cast(DataFrame, result)
+                    # Cast to DataFrame-like type
+                    result = self.obj._constructor(result)
```
this is not the same, is mypy complaining unless you do this?

JBGreisman replied:
yes -- mypy was complaining about this usage:
```python
result = cast(type(self.obj), result)
```
Please let me know if you have an alternative suggestion.

jreback replied:
you can try
`assert isinstance(result, type(self.obj))`

cc @simonjayhawkins 

we don't want to actually call the constructor again

jreback commented on pandas/core/groupby/generic.py:
```diff
@@ -1288,7 +1296,7 @@ def _wrap_applied_output(self, keys, values, not_indexed_same=False):
                     )
                 else:
-                    backup = first_not_none._constructor(**kwargs)
+                    backup = first_not_none._constructor(**kwargs, dtype=object)
```
really you had to add a dtype?

JBGreisman replied:
I had added the dtype to avoid the DeprecationWarning when called with a `Series` (or subclass). You're correct that this is probably unnecessary/wrong -- I'll revert this back.

JBGreisman replied:
I changed the if statement above this line to check if `isinstance(first_not_none, Series)` so that it would also work for subclassed `Series`. This avoids the DeprecationWarnings that were being raised by my tests.

JBGreisman committed:
Revert unnecessary changes in GroupBy()

JBGreisman committed:
Fix test to expect Series from GroupBy.ngroup() and GroupBy.cumcount()

JBGreisman committed:
Fix formatting of groupby.py

JBGreisman committed:
Merge remote-tracking branch 'upstream/master' into groupby-preserve-subclass

JBGreisman commented:
Just to follow-up -- I have a few questions / comments on how best to proceed:

1. The change I made to appease mypy seems to pass all tests, but please let me know if you have a better way to handle that line.
2. Although I initially agreed with having `GroupBy.ngroup()` and `GroupBy.cumcount()` return `Series`, I think it makes sense to have them preserve subclasses (as I initially had it). This would make it so that all `groupby.agg()` / `groupby.transform()` operations return the calling subclass, rather than having two operations that return a different type. 
3. I have one failing build (Linux py37_np_dev), but inspecting the failure messages suggests that it's caused by the `DeprecationWarning` line that I had initially added a `dtype=object` to above. Please let me know if there's anything I should do about this line, or if I should leave it without the explicit dtype. 

Thanks for your help,
Jack

JBGreisman committed:
Avoid DeprecationWarning by checking for instance of Series

JBGreisman committed:
Merge remote-tracking branch 'upstream/master' into groupby-preserve-subclass

jreback commented on pandas/core/groupby/groupby.py:
```diff
@@ -1420,8 +1420,12 @@ def size(self):
 
+        # GH28330 preserve subclassed Series/DataFrames through calls
         if isinstance(self.obj, Series):
+            result = self.obj._constructor(result)
```
try using assert here as well

JBGreisman replied:
Thanks -- this sort of assertion worked to avoid unnecessary calls to the constructor

JBGreisman committed:
Remove unnecessary constructor call in DataFrameGroupBy._cython_agg_blocks()

JBGreisman committed:
Fix mypy static typing issue in DataFrameGroupBy._cython_agg_blocks()

JBGreisman committed:
Ensure consistent return types for GroupBy.size(), ngroup(), and cumcount()

JBGreisman commented:
Thanks for the mypy advice -- I managed to refactor that section a bit to only call `cast()` if needed and avoid calling the constructor again. 

The one additional change made here is to make sure `GroupBy.size()`, `GroupBy.ngroup()`, and `GroupBy.cumcount()` preserve subclassed types. While you had previously mentioned just keeping these return types as `Series`, I think these changes minimize the possibility for surprise when subclassing `DataFrame`/`Series` by ensuring that groupby reductions/transformations preserve the subclasses.

JBGreisman commented:
@jreback -- please let me know if there are any additional changes to this PR that you think are necessary.

jreback commented on pandas/core/groupby/generic.py:
```diff
@@ -1071,7 +1079,10 @@ def _cython_agg_blocks(
                     continue
                 else:
-                    result = cast(DataFrame, result)
+                    # Ensure result is DataFrame-like type
```
can you make this and elif (and include it rather than the else)

JBGreisman replied:
Looking back at this section, my changes were unnecessary. I reverted my changes in `DataFrameGroupBy._cython_agg_blocks()` back to the master branch -- this section is back to the original try-except-else statement.

jreback commented on pandas/core/groupby/groupby.py:
```diff
@@ -2116,7 +2121,13 @@ def ngroup(self, ascending: bool = True):
             index = self._selected_obj.index
-            result = Series(self.grouper.group_info[0], index)
+
+            # GH28330 preserve subclassed Series/DataFrames
```
can you make this more concise e.g.

klass = self.obj_constructor if isinstance(self.obj, Series) else self.obj._constructor_sliced
result = klass(.....)

jreback replied:
also ok would be to add a property to GroupBy which return this e.g.

```
@property
def _constructor(self):
.....
```

JBGreisman replied:
Good idea -- I added a `GroupBy._constructor` property to consolidate the relevant logic in one place. This allowed me to change `ngroup`, `cumcount`, and `size` to use more concise statements.

jreback commented on pandas/core/groupby/groupby.py:
```diff
@@ -2178,7 +2189,12 @@ def cumcount(self, ascending: bool = True):
             cumcounts = self._cumcount_array(ascending=ascending)
-            return Series(cumcounts, index)
+
+            # GH28330 preserve subclassed Series/DataFrames
```
same as above

jreback reviewed (changes requested):
small comment, ping on green.

JBGreisman committed:
Revert DataFrameGroupBy._cython_agg_blocks() back to origin/master

JBGreisman committed:
Add GroupBy._constructor() to facilitate preserving subclassed types

JBGreisman commented:
@jreback I made the recommend changes:
- I reverted the changes in `DataFrameGroupBy._cython_agg_blocks()` because they did not prove necessary to pass my tests.
- I added a `GroupBy._constructor` property to simplify the calls in `ngroup`, `cumcount`, and `size`.

simonjayhawkins commented on pandas/core/groupby/generic.py:
```diff
@@ -1495,7 +1507,9 @@ def _transform_fast(self, result: DataFrame, func_nm: str) -> DataFrame:
             output.append(res)
 
-        return DataFrame._from_arrays(output, columns=result.columns, index=obj.index)
+        return type(self.obj)._from_arrays(
```
why is this not self.obj._constructor like elsewhere?

JBGreisman replied:
Good catch -- I don't think there's a reason for this to be any different. I made the change to use `self.obj._constructor._from_arrays()` for consistency/clarity and it passes all the `groupby` tests.

simonjayhawkins reviewed (commented):
@JBGreisman generally lgtm. one question.

JBGreisman committed:
Change DataFrameGroupBy._transform_fast() to use _constructor property

JBGreisman commented:
@simonjayhawkins Thanks for the comment -- I agree that line should have used `self.obj._constructor`, and changed it

jreback commented on pandas/core/groupby/groupby.py:
```diff
@@ -1182,6 +1182,16 @@ class GroupBy(_GroupBy[FrameOrSeries]):
     """
 
+    @property
+    def _constructor(self) -> Type["Series"]:
```
hmm this should be FrameOrSeries I think.

why is there an else here?

JBGreisman replied:
I had thought this should be a Series because if `self.obj` is a DataFrame, it returns `self.obj._constructor_sliced`. Do you think I should name this property `_series_constructor` or something comparable to make that behavior more apparent?

The else was there because `mypy` was complaining about a missing return statement. I can restructure this with an assertion to avoid an else statement and keep `mypy` from complaining:
```python
@property
def _series_constructor(self) -> Type["Series"]:
    # GH28330 preserve subclassed Series/DataFrames                                                                                                                                                   
    if isinstance(self.obj, DataFrame):
        return self.obj._constructor_sliced
    assert isinstance(self.obj, Series)
    return self.obj._constructor
```
Please let me know if you have a better way to structure this.

jreback replied:
yes pls use an assertion

maybe @simonjayhawkins or @WillAyd can help with the annotation itself

JBGreisman replied:
thanks -- I changed `_constructor` to use the assertion as above. I think it would also make sense to change the name of the property to `_series_constructor` in order to clarify the return type, but I'll hold off for additional comments.

WillAyd replied:
I think the annotation is correct given the way it is implemented, though I am not sure about the implementation. Why do we need to dispatch to `constructor_sliced` for DataFrames? Seems slightly unnatural to have to force that

JBGreisman replied:
This ends up being used in `GroupBy.ngroup()`, `GroupBy.cumcount()`, and `GroupBy.size()`. Prior to these changes, these methods had returned `Series`, regardless of whether they were called from a `SeriesGroupBy` or `DataFrameGroupBy` object. I had updated this to still return a `Series`-type while preserving the subclasses -- to avoid things reverting back to `pd.Series` if they were called from a subclassed DataFrame/Series.

As such, the motivation for dispatching to `constructor_sliced` for DataFrames was to avoid changing the return type for these different `GroupBy` methods from their prior behavior. Do you think it would make sense to make a larger change here that alters these functions to have different return types if called from `SeriesGroupBy` vs. `DataFrameGroupBy`?

WillAyd replied:
Hmm that's unfortunate... Can you rename this property to `obj_constructor` instead? I find the current name a little confusing

WillAyd replied:
Or maybe even `_obj_1d_constructor` to be even more explicit. This is definitely for special cases so want to signal as such

JBGreisman replied:
yeah -- I think `_obj_1d_constructor` is most clear. I'll make the changes.

JBGreisman committed:
Restructure GroupBy._constructor() to remove else statement

JBGreisman committed:
Rename GroupBy._constructor property to GroupBy._obj_1d_constructor

JBGreisman commented:
As per the discussion above, the GroupBy property has been renamed to `_obj_1d_constructor` to clarify its expected return type, and I refactored it to avoid using an `else` statement.

jreback commented:
thanks @JBGreisman very nice, thanks for sticking with it and being responsive!

JBGreisman commented:
no worries -- thanks for the help/suggestions!
